src/commons/editor/Editor.tsx

```diff
@@ -330,23 +331,43 @@ const handlers = {
 
 const EditorBase = React.memo((props: EditorProps & LocalStateProps) => {
   const reactAceRef: React.MutableRefObject<AceEditor | null> = React.useRef(null);
+  const [filePath, setFilePath] = React.useState<string | undefined>(undefined);
 
   // Refs for things that technically shouldn't change... but just in case.
   const handleEditorUpdateBreakpointsRef = React.useRef(props.handleEditorUpdateBreakpoints);
   const handlePromptAutocompleteRef = React.useRef(props.handlePromptAutocomplete);
 
+  const editor = reactAceRef.current?.editor;
+  // Set edit session history when switching to another editor tab.
+  if (editor !== undefined) {
+    if (filePath !== props.filePath) {
+      // Unfortunately, the current editor sets the mode globally.
+      // The side effects make it very hard to ensure that the correct
+      // mode is set for every EditSession. As such, we add this one
+      // line to always propagate the mode whenever we set a new session.
+      // See AceHelper#selectMode for more information.
+      props.session.setMode(editor.getSession().getMode());
```

Comment:
Can't say I'm too happy about this, but in theory, as long as the current mode is always correct, this should work. The proper way would be to not set the mode globally but this hack is too ingrained in the codebase...